### PR TITLE
Fixed issue with maps with nested struct types for sink task

### LINKED ISSUE #507: Nested Maps in Struct are not correctly parsed and lead to later error in writing the sink record
## Description
We currently have a heavily nested avro object that should be written with the SinkConnector to Cosmos DB. This process fails since the in the background used JacksonParser cannot handle the Struct object. The stracktrace looks like:

```
"trace" : "org.apache.kafka.connect.errors.ConnectException: Exiting WorkerSinkTask due to unrecoverable exception.
\tat org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:618)
\tat org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:334)
\tat org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:235)
\tat org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:204)
\tat org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:201)
\tat org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:256)
\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
\tat java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: Failed to serialize the object into json
\tat com.azure.cosmos.implementation.Utils.serializeJsonToByteBuffer(Utils.java:702)
\tat com.azure.cosmos.implementation.InternalObjectNode.serializeJsonToByteBuffer(InternalObjectNode.java:116)
\tat com.azure.cosmos.BridgeInternal.serializeJsonToByteBuffer(BridgeInternal.java:105)
\tat com.azure.cosmos.implementation.RxDocumentClientImpl.getCreateDocumentRequest(RxDocumentClientImpl.java:1508)
\tat com.azure.cosmos.implementation.RxDocumentClientImpl.upsertDocumentInternal(RxDocumentClientImpl.java:1885)
\tat com.azure.cosmos.implementation.RxDocumentClientImpl.lambda$upsertDocument$39(RxDocumentClientImpl.java:1877)
\tat com.azure.cosmos.implementation.ObservableHelper.lambda$inlineIfPossibleAsObs$1(ObservableHelper.java:44)
\tat com.azure.cosmos.implementation.BackoffRetryUtility.lambda$executeRetry$0(BackoffRetryUtility.java:36)
\tat reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:44)
\tat reactor.core.publisher.FluxRetryWhen.subscribe(FluxRetryWhen.java:77)
\tat reactor.core.publisher.MonoRetryWhen.subscribeOrReturn(MonoRetryWhen.java:46)
\tat reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:57)
\tat reactor.core.publisher.MonoDeferContextual.subscribe(MonoDeferContextual.java:55)
\tat reactor.core.publisher.Mono.subscribe(Mono.java:4490)
\tat reactor.core.publisher.Mono.block(Mono.java:1741)
\tat com.azure.cosmos.CosmosContainer.blockItemResponse(CosmosContainer.java:241)
\tat com.azure.cosmos.CosmosContainer.upsertItem(CosmosContainer.java:205)
\tat com.azure.cosmos.kafka.connect.sink.PointWriter.writeCore(PointWriter.java:31)
\tat com.azure.cosmos.kafka.connect.sink.SinkWriterBase.write(SinkWriterBase.java:26)
\tat com.azure.cosmos.kafka.connect.sink.CosmosDBSinkTask.put(CosmosDBSinkTask.java:137)
\tat org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:584)
\t... 10 more
\tSuppressed: java.lang.Exception: #block terminated with an error
\t\tat reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
\t\tat reactor.core.publisher.Mono.block(Mono.java:1742)
\t\t... 16 more
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.apache.kafka.connect.data.Struct and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: java.util.HashMap[\"FIELD\"]->java.util.HashMap[\"FIELD_BELOW\"]->java.util.HashMap[\"5629\"]->java.util.HashMap[\"3450\"])
\tat com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77)
\tat com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1306)
\tat com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:408)
\tat com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.failForEmpty(UnknownSerializer.java:53)
\tat com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.serialize(UnknownSerializer.java:30)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeFields(MapSerializer.java:808)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serializeWithoutTypeInfo(MapSerializer.java:764)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:720)
\tat com.fasterxml.jackson.databind.ser.std.MapSerializer.serialize(MapSerializer.java:35)
\tat com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:480)
\tat com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:319)
\tat com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4624)
\tat com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3828)
\tat com.azure.cosmos.implementation.Utils.serializeJsonToByteBuffer(Utils.java:698)
```
## Possible working Solution (tested)
I could not make a easy way to reproduce this error but i''ll try to provide it as soon as possible. A solution i listed already in another bug report worked though. I changed the StructToJsonMap

```
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.cosmos.kafka.connect.sink;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.kafka.connect.data.Struct;
import org.apache.kafka.connect.json.JsonConverter;

import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class StructToJsonMap {

    public static Map<String, Object> toJsonMap(Struct struct) {
        Map<String, Object> jsonMap = new HashMap<>();

        JsonConverter jsonConverter = new JsonConverter();
        jsonConverter.configure(
                new HashMap<>(){{
                    put("schemas.enable", false);
                    put("decimal.format", "NUMERIC");
                    put("json.date.format", "yyyy-MM-dd'T'HH:mm:ss.SSSXXX");
                }}, false);

        byte[] payload = jsonConverter.fromConnectData("", struct.schema(), struct);
        String json = new String(payload, StandardCharsets.UTF_8);
        ObjectMapper mapper = new ObjectMapper();
        try {
            jsonMap = mapper.readValue(json, Map.class);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }

        return jsonMap;
    }

}
````

Solution looks bit hacky at first but i think its fine and it gives some huge benefits:
- Properties for Converter could be provided from the Connector configuration
- Data written to Cosmos DB could be configured easily for the user (if certain dates are prefered or decimal format)
- Converter offers even some more options
- Complexity in the code is replaced by already well tested code piece

This solution needs better error handling in case the ObjectMapper fails or the return value could just be plain json string (so just remove the Object Mapper). This code did not went through the style check. For testing i removed the plugins in the pom to build the project.

Data itself looks good in Cosmos DB:

```
    "someField": {
        "someField1": {
            "someField2": "eguy0yuv7i",
            "someField3": 4709,
            "someField4": "cx1gpmmnlo",
            "someField5": "gdnrvzmwp5",
            "someField6": "eiw4azchy6",
            "someField7": 1680011226474,
            "someField8": -2598,
            "someField9": {
                "levelLower1": 0.32,
                "levelLower2": 0.2,
                "levelLower3": 0.89,
                "levelLower4": 0.29,
                "levelLower5": 0.69,
                "levelLower6": 0.96
```

I'd really suggest this change since it's a really easy solution and might even fix unknown bugs.

### PR DESCRIPTION
**Pull request by kushagraThapar**
* Fixes #507 
* Fixes security vulnerability in json-path version
* Updated azure-cosmos to 4.42.0

### CODE REVIEW

**Review by maoljaca — DISMISSED** [review]
Looks good. Minor issue might be that handleMap function and toJsonMap handle the List case kinda the same. This could be one major function, but would need a bigger refactoring. This should be sufficient though.

**Review by aayush3011 — APPROVED** [review]
LGTM, thanks @kushagraThapar

File: pom.xml [hunk 1]
```diff
@@ -16,7 +16,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <maven.compiler.source>11</maven.compiler.source>
         <maven.compiler.target>11</maven.compiler.target>
-        <kafka.version>2.8.1</kafka.version>
+        <kafka.version>3.4.0</kafka.version>
```
Comment (kirankumarkolli) [inline pom.xml:19]: Dependency major version raise, will it impact minor expectations?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks for the quick solution :)

File: src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java [hunk 2]
```diff
@@ -73,11 +73,48 @@ public static Map<String, Object> toJsonMap(Struct struct) {
                 case STRUCT:
                     jsonMap.put(fieldName, toJsonMap(struct.getStruct(fieldName)));
                     break;
+                case MAP:
```
Comment (kirankumarkolli) [inline src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java:76]: Any other possible types that needs to be covered?

File: src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java [hunk 3]
```diff
@@ -73,11 +73,48 @@ public static Map<String, Object> toJsonMap(Struct struct) {
...
+        }
+        Map<String, Object> cacheMap = new HashMap<>();
+        map.forEach((key, value) -> {
+            if (value instanceof Map) {
+                cacheMap.put(key, handleMap((Map<String, Object>) value));
+            } else if (value instanceof Struct) {
+                cacheMap.put(key, toJsonMap((Struct) value));
+            } else if (value instanceof List) {
+                List<Object> list = (List<Object>) value;
+                if (list.size() > 0 && list.get(0) instanceof Struct) {
```
Comment (kirankumarkolli) [inline src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java:99]: Is List expected to be of same types for all?

File: pom.xml (hunk 1 above)
Comment (kushagraThapar) [reply pom.xml:19]: Valid concern, I will revert this change. Doesn't make sense to update the major version yet.

File: src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java (hunk 3 above)
Comment (kushagraThapar) [reply src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java:99]: Yes, list should have same types for all data types. However, I have updated the code to check for each item individually.

File: src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java (hunk 2 above)
Comment (kushagraThapar) [reply src/main/java/com/azure/cosmos/kafka/connect/sink/StructToJsonMap.java:76]: Not really,  I think that should cover the cases.

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Pull request merged**